indico/modules/events/controllers/creation.py

```diff
@@ -0,0 +1,125 @@
+
+from __future__ import unicode_literals
+
+from datetime import time, datetime
+
+from dateutil.relativedelta import relativedelta
+from flask import flash, redirect, request, render_template
+from markupsafe import Markup
+from pytz import timezone
+from werkzeug.utils import cached_property
+
+from indico.core.config import Config
+from indico.core.db.sqlalchemy.util.models import get_simple_column_attrs
+from indico.modules.categories import Category
+from indico.modules.events.forms import EventCreationForm, LectureCreationForm
+from indico.modules.events.models.events import EventType
+from indico.modules.events.models.persons import EventPersonLink, EventPerson
+from indico.modules.events.models.series import EventSeries
+from indico.modules.events.operations import create_event
+from indico.util.date_time import now_utc
+from indico.util.struct.iterables import materialize_iterable
+from indico.web.flask.util import url_for
+from indico.web.forms.base import FormDefaults
+from indico.web.util import url_for_index, jsonify_data, jsonify_template
+from MaKaC.webinterface.rh.base import RHProtected
+
+
+class RHCreateEvent(RHProtected):
+    """Create a new event"""
+
+    CSRF_ENABLED = True
+
+    def _checkParams(self):
+        self.event_type = EventType[request.view_args['event_type']]
+
+    @cached_property
+    def _default_category(self):
+        try:
+            category_id = int(request.args['category_id'])
+        except (ValueError, KeyError):
+            return None
+        else:
+            return Category.get(category_id, is_deleted=False)
+
+    def _get_form_defaults(self):
+        category = self._default_category
+        tzinfo = timezone(Config.getInstance().getDefaultTimezone())
+        if category is not None:
+            tzinfo = timezone(category.timezone)
+
+        # try to find good dates/times
+        now = now_utc(exact=False)
+        start_dt = now + relativedelta(hours=1, minute=0)
+        if start_dt.astimezone(tzinfo).time() > time(18):
+            start_dt = tzinfo.localize(datetime.combine(now.date() + relativedelta(days=1), time(9)))
+        end_dt = start_dt + relativedelta(hours=2)
+
+        # XXX: Do not provide a default value for protection_mode. It is selected via JavaScript code
+        # once a category has been selected.
+        return FormDefaults(category=category if category and not category.children else None,
+                            timezone=tzinfo.zone, start_dt=start_dt, end_dt=end_dt,
+                            occurrences=[(start_dt, end_dt - start_dt)],
+                            location_data={'inheriting': False})
+
+    def _create_event(self, data):
+        data = data.copy()
+        return create_event(data.pop('category'), self.event_type, data)
+
+    @materialize_iterable()
+    def _create_series(self, data):
+        @materialize_iterable(dict)
+        def _copy_person_link_data(link_data):
```

Comment:
In event cloning we do it for persons and links separately since we clone a bunch of persons (without adding them to the DB yet) and then the links separately (depending what you actually clone, and this adds the persons that are actually needed to the DB)